### PR TITLE
Add ability to splice stages before a Tail stage

### PR DESCRIPTION
**Pull request by bryce-anderson**
Added tests and fixed the broken spliceBefore on Head stage.

### CODE REVIEW

**Comment by bryce-anderson** [comment]
I need this for the http4s blaze client refactoring work: I want to be able to splice timeout stages into the pipeline on startup.

I plan to merge this rather fast so speak up if you have objections.

File: core/src/main/scala/org/http4s/blaze/pipeline/Stages.scala [hunk 1]
```diff
@@ -314,17 +322,21 @@ trait MidStage[I, O] extends Tail[I] with Head[O] {
...
-    me._nextStage._prevStage = me._prevStage
-
-    _nextStage = null
-    _prevStage = null
-    this
+      _nextStage = null
+      _prevStage = null
+    }
+    else {
+      val e = new Exception("Cannot remove a disconnected stage")
```
Comment (rossabaker) [inline core/src/main/scala/org/http4s/blaze/pipeline/Stages.scala]: I'm not sure this needs to be an exception: the end result is the stage isn't there.  I'd long a stern warning.

**Comment by rossabaker** [comment]
:+1:.  I defer on the exception.

File: core/src/main/scala/org/http4s/blaze/pipeline/Stages.scala (hunk 1 above)
Comment (bryce-anderson) [inline core/src/main/scala/org/http4s/blaze/pipeline/Stages.scala]: Thats a good point. Its now a warning.

**Comment by bryce-anderson** [comment]
Thanks for the review!

**Pull request merged**